src/OpenSpaceToolkit/Astrodynamics/Trajectory/Segment.cpp

```diff
@@ -584,38 +710,233 @@ const NumericalSolver& Segment::accessNumericalSolver() const
-}
+    bool segmentConditionIsSatisfied = eventCondition_->isSatisfied(aState, aState);
 
-Segment::Solution Segment::solveToNextManeuver(const State& aState, const Duration& maximumPropagationDuration) const
-{
-    return solve_(aState, maximumPropagationDuration, false);
+    // Helper lambda to solve a coast segment and update segmentStates
+    const Instant maximumInstant = aState.accessInstant() + maximumPropagationDuration;
+    const auto solveAndAddCoastSegment = [&](const Instant& endInstant) -> bool
+    {
+        const State& lastState = segmentStates.accessLast();
+        const Segment::Solution coastSegmentSolution =
+            solveCoast_(lastState, std::min(endInstant, maximumInstant) - lastState.accessInstant());
+
+        segmentStates.add(Array<State>(coastSegmentSolution.states.begin() + 1, coastSegmentSolution.states.end()));
+
+        segmentConditionIsSatisfied = coastSegmentSolution.conditionIsSatisfied;
+
+        return segmentConditionIsSatisfied;
+    };
+
+    // Helper lambda to solve a single maneuver and extract results
+    const auto solveManeuver = [&](const Shared<Thruster>& thrusterDynamics
+                               ) -> std::pair<Segment::Solution, Array<flightManeuver>>
+    {
+        const State& lastState = segmentStates.accessLast();
+        const Segment::Solution maneuverSolution =
+            solveManeuver_(lastState, maximumInstant - lastState.accessInstant(), thrusterDynamics);
+
+        if (maneuverSolution.states.getSize() <= 2)
+        {
+            return {maneuverSolution, Array<flightManeuver>::Empty()};
+        }
+
+        return {maneuverSolution, maneuverSolution.extractManeuvers(aState.accessFrame())};
+    };
+
+    // Helper lambda to accept a maneuver solution, updating segmentStates and segmentHeterogenousGuidanceLaw
+    // It also updates the last maneuver interval
+    const Shared<HeterogeneousGuidanceLaw> segmentHeterogenousGuidanceLaw =
+        std::make_shared<HeterogeneousGuidanceLaw>();
+    const auto acceptManeuver = [&](const Segment::Solution& maneuverSolution, const Interval& maneuverInterval) -> void
+    {
+        BOOST_LOG_TRIVIAL(debug) << "Maneuver accepted: " << maneuverInterval.toString() << std::endl;
+
+        segmentStates.add(Array<State>(maneuverSolution.states.begin() + 1, maneuverSolution.states.end()));
+
+        segmentConditionIsSatisfied = maneuverSolution.conditionIsSatisfied;
+
+        previousManeuverInterval = maneuverInterval;
+
+        segmentHeterogenousGuidanceLaw->addGuidanceLaw(
+            maneuverSolution.getThrusterDynamics()->getGuidanceLaw(), maneuverInterval
+        );
+    };
+
+    // Helper lambda to handle maximum duration violation strategies, returning true if the segment condition is
+    // satisfied. It also updates the segment states and the guidance law if the maneuver is accepted.
+    const auto handleMaximumDurationViolation = [&](const Interval& candidateManeuverInterval) -> bool
```

Comment:
Same here if you can be explicit with what variables you're capturing in the lambda, since these values get potentially mutated